todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.java

```diff
@@ -0,0 +1,190 @@
+                               @NonNull TasksRepository tasksRepository,
+                               @NonNull TaskDetailNavigator navigator) {
+        mTaskId = taskId;
+        mTasksRepository = checkNotNull(tasksRepository, "tasksRepository cannot be null!");
+        mNavigator = checkNotNull(navigator, "navigator cannot be null");
+        mLoadingSubject = BehaviorSubject.create(false);
+        mSnackbarText = PublishSubject.create();
+    }
+
+    /**
+     * @return a stream notifying on whether the loading is in progress or not
+     */
+    @NonNull
+    public Observable<Boolean> getLoadingIndicatorVisibility() {
+        return mLoadingSubject.asObservable();
+    }
+
+    /**
+     * @return a stream containing the task model. An error will be emitted
+     * if the task id is invalid. The loading is updated before retrieving the task and when the
+     * task has been retrieved.
+     */
+    @NonNull
+    public Observable<TaskUiModel> getTaskUiModel() {
+        if (Strings.isNullOrEmpty(mTaskId)) {
+            return Observable.error(new Exception("Task id null or empty"));
+        }
+
+        return mTasksRepository.getTask(mTaskId)
+                .map(this::createModel)
+                .doOnSubscribe(() -> mLoadingSubject.onNext(true))
+                .doOnNext(__ -> mLoadingSubject.onNext(false));
+    }
+
+    @NonNull
+    private TaskUiModel createModel(Task task) {
+        boolean isTitleVisible = !Strings.isNullOrEmpty(task.getTitle());
+        boolean isDescriptionVisible = !Strings.isNullOrEmpty(task.getDescription());
+
+        return new TaskUiModel(isTitleVisible, task.getTitle(), isDescriptionVisible,
+                task.getDescription(), task.isCompleted());
+    }
+
+    /**
+     * @return a stream that emits when a snackbar should be displayed. The stream contains the
+     * snackbar text
+     */
+    @NonNull
+    public Observable<Integer> getSnackbarText() {
+        return mSnackbarText.asObservable();
+    }
+
+    /**
+     * Handle the response received on onActivityResult.
+     *
+     * @param requestCode the request with which the Activity was opened.
+     * @param resultCode  the result of the Activity.
+     */
+    public void handleActivityResult(int requestCode, int resultCode) {
+        if (TaskDetailActivity.REQUEST_EDIT_TASK == requestCode
+                && Activity.RESULT_OK == resultCode) {
+            mNavigator.onTaskEdited();
+        }
+    }
+
+    /**
+     * @return a stream that completes when the task was edited.
+     */
+    @NonNull
+    public Completable editTask() {
+        return Completable.fromAction(() -> {
+            if (Strings.isNullOrEmpty(mTaskId)) {
+                throw new RuntimeException("Task id null or empty");
+            }
+            mNavigator.onStartEditTask(mTaskId);
+        });
+    }
+
+    /**
+     * Deletes the task from repository. Emits when the task has been deleted. An error is emitted
```

Comment:
I would split this comment and one below that into three sentences where every would start at the new line like in `TaskId` is already done.